man/supl-cert.1
.\"EMACS: -*- nroff -*-

.TH SUPL-CERT 1 "version 1.0"
.SH NAME
supl-cert \- utility to create SSL keys and certificates for supl-proxy
.SH SYNOPISIS
.B supl-cert
[\fIdoman-name\fP]
.br
.SH DESCRIPTION
\fBsupl-cert\fP takes as the only parameter the domain name of the
server where the supl-proxy will be ran and what is set to client
(e.g. mobile phone) as the SUPL server.

.SH OPTIONS
The \fIdomain-name\fP of the supl-proxy machine.
.SH NOTES
.B *** DO NOT USE THESE KEYS FOR ANYTHING REQUIRING REAL SECURITY ***
.br

\fBsupl-proxy\fP expects to see these certificate and key files with
these names in the current working directory.

The CA certificate \fIca-cert.pem\fP file must be given to the SUPL
client device (e.g. mobile phone).
.SH FILES
.TP
.I ca-cert.pem
CA certificate. Copy over to the SUPL client device.
.TP
.I ca-priv.pem
CA private key. This file is not really needed.
.TP
.I srv-cert.pem
supl-proxy server certificate, signed by the CA certificate.
.TP
.I srv-priv.pem
supl-proxy server private key.
.SH BUGS
Please send any comments or bug reports to \fBtatu -at- tajuma.com\fP.
.SH HOMEPAGE
http://www.tajuma.com/supl
.SH AUTHOR
Tatu Männistö <tatu -at- tajuma.com>


debian/changelog
supl (1.0.6) unstable; urgency=low

  * Fixed BUG#3427808 again...

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 27 Oct 2011 21:55:00 -0400

supl (1.0.5) unstable; urgency=low

  * Fixed BUG#3427808, symlinks is asn1-tarballs

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 22 Oct 2011 22:45:00 -0400

supl (1.0.4) unstable; urgency=low

  * Fixed failing to decode too long messages
  * Added proper scaling factors for --format human

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 22 Oct 2011 22:45:00 -0400

supl (1.0.3) unstable; urgency=low

  * Update precompiled asn1c files using asn1c version 0.9.23
  * Fix README to reflect these changes, fix few other bugs there too
  * Add asn1 and supl debugging configuration options to configure

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 20 Oct 2011 22:15:00 -0400

supl (1.0.2) unstable; urgency=low

  * Bug#3425862 Fixes command line option parsing bug

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 19 Oct 2011 18:02:00 -0400

supl (1.0.1) unstable; urgency=low

  * Fixes reading ephemeris for all satellites
  * Add -a flag to request almanac data from the server

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 17 Oct 2011 22:41:00 -0400

supl (1.0) unstable; urgency=low

  * Fixes multiple build issues, warnings, etc.
  * Uses now plain Makefiles instead of autotools.
  * Got rid of cert utilities and combined them into supl-cert.
  * Manual pages added.

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 28 Mar 2011 16:54:00 -0400

supl (0.9.2) unstable; urgency=low

  * Uses slpSessionID if given by the supl server.
  * Uses wcdma and gsm cell info.
  * Fixed few bugs all round.

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Wed, 2 Mar 2011 16:02:00 -0400

supl (0.9.1) unstable; urgency=low

  * Added almanac and acquisition assistance support.
  * Fixed few build errors.

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Mon, 1 Oct 2007 22:02:00 -0400

supl (0.9) unstable; urgency=low

  * First semi-public release.

 -- Tatu Mannisto <tatu -at- tajuma -dot- com>  Sun, 30 Sep 2007 14:15:00 -0400



src/supl-cert.c
/* crude hack to get ca and server keys and certificates done */
/* based on example code from OpenSSL */

#include <stdio.h>
#include <stdlib.h>

#include <openssl/pem.h>
#include <openssl/conf.h>
#include <openssl/x509v3.h>

static char *domain;

static int mkit_ca(X509 **x509p, EVP_PKEY **pkeyp, int bits, int serial, int days);
static int mkit_srv(X509 **x509p, EVP_PKEY **pkeyp, EVP_PKEY *sign_key, X509 *sign_cert, int bits, int serial, int days);

int main(int argc, char *argv[])
	{
	BIO *bio_err;
	X509 *x509=0;
	X509 *sign_cert=0;
	EVP_PKEY *pkey=0;
	EVP_PKEY *sign_key=0;

	if (argc != 2) {
	  fprintf(stderr, "usage: %s domain-name\n", argv[0]);
	  exit(1);
	}
	domain = argv[1];

	CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);

	bio_err=BIO_new_fp(stderr, BIO_NOCLOSE);

	mkit_ca(&x509,&pkey,1024,42,365);

#if 0
	RSA_print_fp(stdout,pkey->pkey.rsa,0);
	X509_print_fp(stdout,x509);
#endif

	{
	  FILE *f;
	  f = fopen("ca-priv.pem", "w");
	  PEM_write_PrivateKey(f,pkey,NULL,NULL,0,NULL, NULL);
	  fclose(f);
	}
	{
	  FILE *f;
	  f = fopen("ca-cert.pem", "w");
	  PEM_write_X509(f, x509);
	  fclose(f);
	}

	{
	  FILE *f;
	  f = fopen("ca-priv.pem", "r");
	  if (!f) {
	    fprintf(stderr, "Can not open signature key file ca-priv.pem\n");
	    exit(1);
	  }
	  sign_key = PEM_read_PrivateKey(f, 0, 0, 0);
	  fclose(f);
	  if (!sign_key) {
	    fprintf(stderr, "Signature key not good\n");
	    exit(1);
	  }
	}

	{
	  FILE *f;
	  f = fopen("ca-cert.pem", "r");
	  if (!f) {
	    fprintf(stderr, "Can not open CA certificate file ca-priv.pem\n");
	    exit(1);
	  }
	  sign_cert = PEM_read_X509(f, 0, 0, 0);
	  fclose(f);
	  if (!sign_cert) {
	    fprintf(stderr, "CA cert not good\n");
	    exit(1);
	  }
	}

	X509_free(x509);
	EVP_PKEY_free(pkey);

	x509 = 0;
	pkey = 0;

	mkit_srv(&x509,&pkey,sign_key,sign_cert,1024,0xbeef,365);

#if 0
	RSA_print_fp(stdout,pkey->pkey.rsa,0);
	X509_print_fp(stdout,x509);
#endif

	{
	  FILE *f;
	  f = fopen("srv-priv.pem", "w");
	  PEM_write_PrivateKey(f,pkey,NULL,NULL,0,NULL, NULL);
	  fclose(f);
	}
	{
	  FILE *f;
	  f = fopen("srv-cert.pem", "w");
	  PEM_write_X509(f, x509);
	  fclose(f);
	}

	X509_free(x509);
	EVP_PKEY_free(pkey);

	/* Only needed if we add objects or custom extensions */
	X509V3_EXT_cleanup();
	OBJ_cleanup();

	CRYPTO_mem_leaks(bio_err);
	BIO_free(bio_err);
	return(0);
	}

static void callback(p, n, arg)
int p;
int n;
void *arg;
	{
	return;
	}

static int mkit_ca(x509p,pkeyp,bits,serial,days)
X509 **x509p;
EVP_PKEY **pkeyp;
int bits;
int serial;
int days;
	{
	X509 *x;
	EVP_PKEY *pk;
	RSA *rsa;
	X509_NAME *name=NULL;
	X509_EXTENSION *ex=NULL;

	
	if ((pkeyp == NULL) || (*pkeyp == NULL))
		{
		if ((pk=EVP_PKEY_new()) == NULL)
			{
			abort(); 
			return(0);
			}
		}
	else
		pk= *pkeyp;

	if ((x509p == NULL) || (*x509p == NULL))
		{
		if ((x=X509_new()) == NULL)
			goto err;
		}
	else
		x= *x509p;

	rsa=RSA_generate_key(bits,RSA_F4,callback,NULL);
	if (!EVP_PKEY_assign_RSA(pk,rsa))
		{
		abort();
		goto err;
		}
	rsa=NULL;

	X509_set_version(x,2);
	ASN1_INTEGER_set(X509_get_serialNumber(x),serial);
	X509_gmtime_adj(X509_get_notBefore(x),(long)-60*60*24*31);
	X509_gmtime_adj(X509_get_notAfter(x),(long)60*60*24*days);
	X509_set_pubkey(x,pk);

	name=X509_get_subject_name(x);

	/* This function creates and adds the entry, working out the
	 * correct string type and performing checks on its length.
	 * Normally we'd check the return value for errors...
	 */
	X509_NAME_add_entry_by_txt(name,"C",
				   MBSTRING_ASC, (const unsigned char *)"FI", -1, -1, 0);
	X509_NAME_add_entry_by_txt(name,"O",
				   MBSTRING_ASC, (const unsigned char *)"SuperDupper CA", -1, -1, 0);
	X509_NAME_add_entry_by_txt(name,"CN",
				   MBSTRING_ASC, (const unsigned char *)domain, -1, -1, 0);

	X509_set_issuer_name(x,name);

	/* Add extension using V3 code: we can set the config file as NULL
	 * because we wont reference any other sections. We can also set
         * the context to NULL because none of these extensions below will need
	 * to access it.
	 */

	ex = X509V3_EXT_conf_nid(NULL, NULL, NID_netscape_ssl_server_name, domain);
	X509_add_ext(x,ex,-1);
	X509_EXTENSION_free(ex);

	ex = X509V3_EXT_conf_nid(NULL, NULL, NID_netscape_cert_type,
				 "sslCA");
	X509_add_ext(x,ex,-1);
	X509_EXTENSION_free(ex);

	ex = X509V3_EXT_conf_nid(NULL, NULL, NID_basic_constraints,
				 "critical,CA:TRUE");
	X509_add_ext(x,ex,-1);
	X509_EXTENSION_free(ex);

	ex = X509V3_EXT_conf_nid(NULL, NULL, NID_key_usage, 
				 "critical,keyCertSign,cRLSign");
	X509_add_ext(x,ex,-1);
	X509_EXTENSION_free(ex);

	if (!X509_sign(x,pk,EVP_sha1()))
		goto err;

	*x509p=x;
	*pkeyp=pk;
	return(1);
err:
	return(0);
	}

static int mkit_srv(x509p,pkeyp,sign_key,sign_cert,bits,serial,days)
X509 **x509p;
EVP_PKEY **pkeyp;
EVP_PKEY *sign_key;
X509 *sign_cert;
int bits;
int serial;
int days;
	{
	X509 *x;
	EVP_PKEY *pk;
	RSA *rsa;
	X509_NAME *name=NULL;
	X509_EXTENSION *ex=NULL;

	
	if ((pkeyp == NULL) || (*pkeyp == NULL))
		{
		if ((pk=EVP_PKEY_new()) == NULL)
			{
			abort(); 
			return(0);
			}
		}
	else
		pk= *pkeyp;

	if ((x509p == NULL) || (*x509p == NULL))
		{
		if ((x=X509_new()) == NULL)
			goto err;
		}
	else
		x= *x509p;

	rsa=RSA_generate_key(bits,RSA_F4,callback,NULL);
	if (!EVP_PKEY_assign_RSA(pk,rsa))
		{
		abort();
		goto err;
		}
	rsa=NULL;

	X509_set_version(x,2);
	ASN1_INTEGER_set(X509_get_serialNumber(x),serial);
	X509_gmtime_adj(X509_get_notBefore(x),(long)-60*60*24*31);
	X509_gmtime_adj(X509_get_notAfter(x),(long)60*60*24*days);
	X509_set_pubkey(x,pk);

	name=X509_get_subject_name(x);

	/* This function creates and adds the entry, working out the
	 * correct string type and performing checks on its length.
	 * Normally we'd check the return value for errors...
	 */
	X509_NAME_add_entry_by_txt(name,"C",
				   MBSTRING_ASC, (const unsigned char *) "FI", -1, -1, 0);
	X509_NAME_add_entry_by_txt(name,"CN",
				   MBSTRING_ASC, (const unsigned char *)domain, -1, -1, 0);

	name = X509_get_subject_name(sign_cert);
	X509_set_issuer_name(x,name);

	/* Add extension using V3 code: we can set the config file as NULL
	 * because we wont reference any other sections. We can also set
         * the context to NULL because none of these extensions below will need
	 * to access it.
	 */

	ex = X509V3_EXT_conf_nid(NULL, NULL, NID_key_usage, 
				 "critical,keyEncipherment,dataEncipherment,keyAgreement,nonRepudiation");
	X509_add_ext(x,ex,-1);
	X509_EXTENSION_free(ex);

	if (!X509_sign(x,sign_key,EVP_sha1()))
		goto err;

	*x509p=x;
	*pkeyp=pk;
	return(1);
err:
	return(0);
	}
